### DISCUSSION TITLE
Lifetime of intercept

### DISCUSSION THREAD

**Discussion by Cito** [cypress-io/cypress#9395] [Questions and Help] [answered]
The documentation for `intercept()` should explain the following things:

* How long is the lifetime of an `intercept()` command? (It should be documented that the effect is reset before each test.)
* What happens when you execute multiple `intercept()` commands? Does the second one replace the first? Or are they "stacked" on each other?
* Is it possible to manually remove the effect of all `intercept()` commands within a test (or the effect of the last one if they are "stacked")?

Here are two use cases that I had in mind:

* Having an `intercept()` in `beforeEach()` that should not be effective or altered just for one or a few of the tests.
* I want to simulate a temporary server error.
  - Create an `intercept()` that returns a status code of 500.
  - Try to execute a function in the UI that needs the server, e.g. a login.
  - Check that the UI reacts properly and shows a nice error message to the user.
  - "Repair" the server by removing the effect of the last `intercept()`
  - Try to execute the same function.
  - Check that the UI now does what is expected and there is no more error message.

For the second use case, I could create a dynamic intercept that only returns the error on the first try and otherwise passes the real result from the server. But it seems more complicated to me.

**Comment by Cito** [comment]
Here is the solution with the dynamic intercept.

The following function can be used to simulate a given number of server errors (by default, one error only).

```js
Cypress.Commands.add('simulateNetworkError', (calls = 1, statusCode = 500) => {
  cy.intercept(
    { method: 'POST', url: '/graphql' },
    (req) => {
      if (calls) {
        req.reply({ statusCode });
        --calls;
      }
    }
  );
});
```

**Comment by alycda** [comment]
> (It should be documented that the effect is reset before each test.)

How did you figure this out? I now know why my test set-up wasn't working as expected

**Reply by flotwig** [reply]
It's documented here: https://docs.cypress.io/api/commands/intercept.html#Usage

**Comment by flotwig** [comment, marked as answer]
> How long is the lifetime of an intercept() command? (It should be documented that the effect is reset before each test.)

After each test - https://docs.cypress.io/api/commands/intercept.html#Usage

> What happens when you execute multiple intercept() commands? Does the second one replace the first? Or are they "stacked" on each other?

See discussion on this issue: https://github.com/cypress-io/cypress/issues/9302

> Is it possible to manually remove the effect of all intercept() commands within a test (or the effect of the last one if they are "stacked")?

Not currently, but this is something that would be good to add in the future for advanced use cases.

**Comment by jeremydorne** [comment]
Is there a good approach to reducing the time between an intercept getting destroyed and a new intercept getting created? 

If an entire mock service is setup through intercept, there are a fair amount of instances where an application can make requests to the mocked service between when a given test clears its `intercept`s and the next test creates its `intercept`s. I've seen the guidance describe [here](https://glebbahmutov.com/blog/cypress-intercept-problems/#the-intercept-was-registered-too-late) to wait for intercepts to register but I'm not sure how to apply that logic at scale.

The issue I'm seeing roughly looks like:
```
describe("this suite relies on a mock service", () => {
    beforeEach(() => {
        // setup intercept service
    });

    it("tests something", () => {
        // perform some actions
    });

    // request to intercept service fails

    it("tests something else", () => {
        // perform some actions
    });
}
```

**Comment by the0neWhoKnocks** [comment]
Came to dig through the discussions because I too was confused about the lifecycle of `intercept`. Mainly I just wanted to intercept, mock response once, and destroy the intercept so all other requests behave as normal. Perhaps a chained or new method could solve this common scenario
```js
intercept.once()
// or
interceptOnce()
```

**Reply by flotwig** [reply]
Like this? https://github.com/cypress-io/cypress/issues/8531

**Reply by the0neWhoKnocks** [reply]
Yarp, but since `route` is deprecated, it looks like this link is more appropriate https://github.com/cypress-io/cypress/pull/14513
